numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -1086,6 +1086,95 @@ TIMEDELTA_setitem(PyObject *op, void *ov, void *vap)
 
 /* Assumes contiguous, and aligned, from and to */
 
+/* special case TIMEDELTA NPY_DATETIME_NAT handling */
+
+/**begin repeat
+ *
+ * #TYPE = TIMEDELTA#
+ * #type = npy_timedelta#
+ */
+
+static void
+@TYPE@_to_HALF(void *input, void *output, npy_intp n,
+        void *NPY_UNUSED(aip), void *NPY_UNUSED(aop))
+{
+    const @type@ *ip = input;
+    npy_half *op = output;
+
+    while (n--) {
+        if (*ip == NPY_DATETIME_NAT) {
+            *op = NPY_HALF_NAN;
+        }
+        else {
+            *op = npy_float_to_half((float)(*ip));
+        }
+        ip++;
+        op++;
+    }
+}
+
+static void
+HALF_to_@TYPE@(void *input, void *output, npy_intp n,
+        void *NPY_UNUSED(aip), void *NPY_UNUSED(aop))
+{
+    const npy_half *ip = input;
+    @type@ *op = output;
+
+    while (n--) {
+        *op++ = (@type@)npy_half_to_float(*ip++);
+    }
+}
+
+/**end repeat**/
+#if NPY_SIZEOF_SHORT == 2
+#define HALF_to_HALF SHORT_to_SHORT
+#elif NPY_SIZEOF_INT == 2
+#define HALF_to_HALF INT_to_INT
+#endif
+
+/**begin repeat
+ *
+ * #TOTYPE = BYTE, UBYTE, SHORT, USHORT, INT, UINT, LONG, ULONG,
+ *           LONGLONG, ULONGLONG, FLOAT, DOUBLE, LONGDOUBLE, DATETIME,
+ *           TIMEDELTA#
+ * #totype = npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int, npy_uint,
+ *           npy_long, npy_ulong, npy_longlong, npy_ulonglong,
+ *           npy_float, npy_double, npy_longdouble,
+ *           npy_datetime, npy_timedelta#
```

Comment:
This is a bit lazy on my part -- if we really want to optimize for performance I think we'll need one set of templates that disclude floating point types & avoid the NaT check, and then another for timedelta-> floating types that does have the inner loop NaT check.

The question is whether the code duplication / specialization is worth the performance advantage of avoiding the NaT check on the bystander types here.